### ISSUE TITLE
1.0 suggestion: use functions instead of prepared statements

### ISSUE THREAD

**Issue by bgentry** [chanks/que#176]
Hey there, just had an idea for 1.0. Maybe you've already thought of this, but...

The `sql.rb` file contains some ~large-ish SQL statements which have to be prepared on each client prior to execution (though you don't do this within a transaction to avoid fouling up the txn). I noticed in the v4 migration file you're creating a function for notifying idle workers about jobs they should lock. I like the way you've done this, especially with unlogged tables to track idle workers. Smart 😄 

It got me thinking, would Que be better off creating functions for all of those large SQL statements? You can still version them through migrations. Then you wouldn't have to deal with prepared statements at all and could just call the functions that have already been created. Seems like it should improve performance and decrease the complexity of client libs. One of the annoyances I ran into with que-go was dealing with prepared statements, so it would be cool to see that disappear in the new version.

I'm sure this will make development and versioning a bit more complicated (likely with more schema migrations) but it seems like that tradeoff might be worthwhile. Curious to hear your thoughts though!

Cheers 🍻 ✌️

**Comment by chanks** [comment]
Yeah, the reason I've avoided doing this in the past is the increased number of migrations. There's also the search_path question, for users that might have multiple queues in different schemas - prepared statements take the search_path into account, while functions won't. We use schemas for multitenancy at my day job, and even though we're sticking with one job table in the public schema for now, I've been wondering whether that will change... :thinking: 

It's worth ruminating on, at the very least.
